pkg/v1/tarball/layer.go

```diff
@@ -101,20 +126,38 @@ func LayerFromOpener(opener Opener, opts ...LayerOption) (v1.Layer, error) {
 	}
 
 	layer := &layer{
-		compressed:  compressed,
 		compression: gzip.BestSpeed,
-		opener:      opener,
+	}
+
+	if compressed {
```

Comment:
I'd really like to drop the above (L123) use of `IsGzipped` because this came up recently and is really annoying: https://github.com/google/go-containerregistry/issues/112#issuecomment-735444758

We should be able to call opener just once, which would fix this... what about something like this:

Use a `bufio.Reader` to [`Peek`](https://golang.org/pkg/bufio/#Reader.Peek) at the first 2 bytes. We technically only need a 2 byte buffer for this, but we should probably do 4K to read (at least) a whole block. This will let us know if it's compressed without consuming the reader.

`crane append` uses `tarball.LayerFromFile`, which we should usually be able to call multiple times, but that doesn't seem to work with process substitution without fiddling.

Reply:
Okay, so I went ahead and fixed the thing that I actually care about: https://github.com/google/go-containerregistry/pull/868

I _still_ think it's a good idea to use a `bufio.Reader` here to peek at stuff, but I care about 15% less about it than I did before.